### PR TITLE
[mir-opt] Introduce a new flag to enable experimental/unsound mir opts

### PR DESCRIPTION
**Pull request by wesleywiser**
This implements part of https://github.com/rust-lang/compiler-team/issues/319. The exact name of this flag was not decided as part of that MCP and some people expressed that it should include "unsound" in some way. 

I've chosen to use `enable-experimental-unsound-mir-opts` as the name. While long, I don't think that matters too much as really it will only be used by some mir-opt tests. If you object or have a better name, please leave a comment!

r? @oli-obk 
cc @rust-lang/wg-mir-opt @RalfJung

### CODE REVIEW

**Comment by Aaron1011** [comment]
I would personally prefer `enable-unsound-mir-opts`. It's easier to type, and we could theoretically move a non-experimental optimization pass behind this flag if we discovered it was unsound.

File: src/test/mir-opt/inline/inline-any-operand.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// compile-flags: -Z span_free_formats
+// compile-flags: -Z span_free_formats -Zenable-experimental-unsound-mir-opts
```
Comment (oli-obk) [inline src/test/mir-opt/inline/inline-any-operand.rs:1]: why is this needed? From the test description it doesn't look like this needs copy prop or simplify_try

File: src/test/mir-opt/inline/inline-closure-borrows-arg.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// compile-flags: -Z span_free_formats
+// compile-flags: -Z span_free_formats -Zenable-experimental-unsound-mir-opts
```
Comment (oli-obk) [inline src/test/mir-opt/inline/inline-closure-borrows-arg.rs:1]: same here

File: src/test/mir-opt/inline/inline-trait-method_2.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// compile-flags: -Z span_free_formats -Z mir-opt-level=3
+// compile-flags: -Z span_free_formats -Z mir-opt-level=3 -Zenable-experimental-unsound-mir-opts
```
Comment (oli-obk) [inline src/test/mir-opt/inline/inline-trait-method_2.rs:1]: and here

File: src/test/mir-opt/nrvo-simple.rs [hunk 4]
```diff
@@ -1,3 +1,4 @@
+// compile-flags: -Zenable-experimental-unsound-mir-opts
```
Comment (oli-obk) [inline src/test/mir-opt/nrvo-simple.rs:1]: and here

**Comment by wesleywiser** [comment]
@oli-obk I tried to keep our MIR opt tests' output as close to the current ones as possible. Some of these passes run after copy-prop so the effects of it show being disabled by default (even under mir-opt-level=2) show up as changes in the tests. 

If you think we should not add this flag to those tests and accept the changes instead, I'd be happy to do so.

**Comment by oli-obk** [comment]
> If you think we should not add this flag to those tests and accept the changes instead, I'd be happy to do so.

Maybe remove the flag again in a separate commit so we keep the changes separate?

**Comment by oli-obk** [comment]
With that and the flag renamed to `unsound-mir-opts` as suggested: r=me

**Comment by RalfJung** [comment]
> // FIXME(76740): This opt is buggy and can cause unsoundness.

Is "opt" short for option or optimization? It is used for option in the next line, so probably better to spell it out here.

**Comment by wesleywiser** [comment]
@RalfJung I remember you felt quite strongly the option name should include "unsound". Does `unsound-mir-opts` work for you?

**Comment by tmiasko** [comment]
Is this option intended only for unsound passes or also that are buggy in other ways e.g., ICE? If the latter I would also include inlining pass. Recently when I evaluated mir-opt-level=2 on a different crates, rustc generally ICEd due to normalization issues in inliner or other issues exposed indirectly by inlining. The attempt to enable inlining in #75495 didn't fare any better.

**Comment by wesleywiser** [comment]
Originally, I was thinking this flag could handle both but now I'm thinking we should probably have two. The optimizations that have bugs which result in unsoundness are far worse than the ones that "just" ICE the compiler.

**Comment by RalfJung** [comment]
> @RalfJung I remember you felt quite strongly the option name should include "unsound". Does `unsound-mir-opts` work for you?

Yes I like that name. :)

**Comment by Dylan-DPC** [comment]
failed in [rollup CI](https://github.com/rust-lang/rust/pull/77061/checks?check_run_id=1150431258)

@bors r-

**Comment by wesleywiser** [comment]
Rebased and resolved the failing test from the rollup.

@bors r=oli-obk

**Comment by wesleywiser** [comment]
Rebased

@bors r=oli-obk
